pkg/driver/credential.go

```diff
@@ -0,0 +1,187 @@
+	}
+}
+
+func (c *credentialProvider) provideFromDriver() (*MountCredentials, error) {
+	klog.V(4).Infof("NodePublishVolume: Using driver identity")
+
+	hostPluginDir := hostPluginDirWithDefault()
+	hostTokenPath := path.Join(hostPluginDir, "token")
+
+	return &MountCredentials{
+		AccessKeyID:     os.Getenv(keyIdEnv),
+		SecretAccessKey: os.Getenv(accessKeyEnv),
+		SessionToken:    os.Getenv(sessionTokenEnv),
+		Region:          os.Getenv(regionEnv),
+		DefaultRegion:   os.Getenv(defaultRegionEnv),
+		WebTokenPath:    hostTokenPath,
+		StsEndpoints:    os.Getenv(stsEndpointsEnv),
+		AwsRoleArn:      os.Getenv(roleArnEnv),
+	}, nil
+}
+
+func (c *credentialProvider) provideFromPod(ctx context.Context, req *csi.NodePublishVolumeRequest) (*MountCredentials, error) {
+	klog.V(4).Infof("NodePublishVolume: Using pod identity")
+
+	tokensJson := req.VolumeContext[volumeCtxServiceAccountTokens]
+	if tokensJson == "" {
+		klog.Error("`authenticationSource` configured to `pod` but no service account tokens are received. Please make sure to enable `podInfoOnMountCompat`, see TODO")
+		return nil, status.Error(codes.InvalidArgument, "Missing service account tokens")
+	}
+
+	var tokens map[string]*Token
+	if err := json.Unmarshal([]byte(tokensJson), &tokens); err != nil {
+		return nil, status.Errorf(codes.InvalidArgument, "Failed to parse service account tokens: %v", err)
+	}
+
+	stsToken := tokens[serviceAccountTokenAudienceSTS]
+	if stsToken == nil {
+		klog.Errorf("`authenticationSource` configured to `pod` but no service account tokens for %s received. Please make sure to enable `podInfoOnMountCompat`, see TODO", serviceAccountTokenAudienceSTS)
+		return nil, status.Errorf(codes.InvalidArgument, "Missing service account token for %s", serviceAccountTokenAudienceSTS)
+	}
+
+	volumeID := req.GetVolumeId()
+
+	// TODO: Cleanup these files on unmount and startup.
+	// TODO: Should we make the write atomic by writing to a temporary path and renaming afterwards?
+	err := os.WriteFile(path.Join("/csi/", volumeID+".token"), []byte(stsToken.Token), 0400)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "Failed to write service account token: %v", err)
+	}
+
+	hostPluginDir := hostPluginDirWithDefault()
+	hostTokenPath := path.Join(hostPluginDir, volumeID+".token")
+
+	awsRoleARN, err := c.findPodServiceAccountRole(ctx, req)
+	if err != nil {
+		return nil, err
+	}
+
+	region := os.Getenv(regionEnv)
+	defaultRegion := os.Getenv(defaultRegionEnv)
+
+	// TODO: How to handle missing region? We should ideally try the following:
+	//	1. Region set by users explicitly
+	//	2. Env variable
+	//	3. From IMDS
+	//	4. Fail
+	if region == "" {
+		if defaultRegion != "" {
+			region = defaultRegion
+		} else {
+			region, err = regionFromIMDS()
+			if err != nil {
+				return nil, status.Errorf(codes.InvalidArgument, "Failed to detect AWS Region using IMDS: %v, please explicitly set AWS Region", err)
```

Comment:
Might be worth saying how to do this

Reply:
Once we decide how we want to expose this option, I'll add this to our documentation and link it from here